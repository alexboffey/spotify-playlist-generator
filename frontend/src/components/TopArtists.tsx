import React, { useState } from "react";
import { Query } from "react-apollo";
import { Avatar, message, List, Form, Select } from "antd";
import { startCase } from "lodash";

import Header from "./Header";
import {
  IMyTopArtistsQuery,
  MY_TOP_ARTISTS_QUERY
} from "../graphql/myTopArtistsQuery";

const TopArtists: React.FunctionComponent = () => {
  const initialTimeRange = "short_term";
  const [timeRange, setTimeRange] = useState<string>(initialTimeRange);

  return (
    <Query<IMyTopArtistsQuery>
      query={MY_TOP_ARTISTS_QUERY}
      variables={{ limit: 50, time_range: timeRange }}
      notifyOnNetworkStatusChange
    >
      {({ data, loading, error, refetch }) => {
        if (error) return message.error(error);

        return (
          <React.Fragment>
            <Header title="Top Artists" />
            <Form>
              <Form.Item label="Time Range">
                <Select
                  defaultValue={initialTimeRange}
                  onChange={(value: string) => {
                    setTimeRange(value);
                    refetch();
                  }}
                >
                  <Select.Option value="short_term">Short Term</Select.Option>
                  <Select.Option value="medium_term">Medium Term</Select.Option>
                  <Select.Option value="long_term">Long Term</Select.Option>
                </Select>
              </Form.Item>
            </Form>

            <List loading={loading}>
              {data &&
                !loading &&
                data.myTopArtists.items.map(({ id, name, images, genres }) => (
                  <List.Item key={id}>
                    <List.Item.Meta
                      title={name}
                      description={
                        genres.length > 0
                          ? genres
                              .slice(0, 3)
                              .map(genre => startCase(genre))
                              .join(", ")
                          : "No genres available for this artist"
                      }
                      avatar={
                        <Avatar
                          shape="square"
                          size="large"
                          src={images[0] ? images[0].url : ""}
                        >
                          {!images[0] && name.split("")[0]}
                        </Avatar>
                      }
                    />
                  </List.Item>
                ))}
            </List>
          </React.Fragment>
        );
      }}
    </Query>
  );
};

export default TopArtists;
