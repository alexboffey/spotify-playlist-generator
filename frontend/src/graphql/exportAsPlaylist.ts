import gql from "graphql-tag";

export const EXPORT_AS_PLAYLIST_MUTATION = gql`
  mutation EXPORT_AS_PLAYLIST_MUTATION(
    $playlistName: String
    $tracks: [String]!
  ) {
    exportAsPlaylist(playlistName: $playlistName, tracks: $tracks) {
      name
    }
  }
`;

export interface IExportAsPlaylistMutation {
  exportAsPlaylist: {
    name: string;
  };
}
