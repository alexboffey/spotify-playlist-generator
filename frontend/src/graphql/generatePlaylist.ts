import gql from "graphql-tag";

export const GENERATE_PLAYLIST_QUERY = gql`
  query GENERATE_PLAYLIST_QUERY($seeds: String) {
    generatePlaylist(seeds: $seeds) {
      tracks {
        id
        name
        preview_url
        artists {
          id
          name
        }
        album {
          id
          name
          images {
            url
          }
        }
      }
    }
  }
`;

export interface IGeneratePlaylistQuery {
  generatePlaylist: {
    tracks: Array<{
      id: string;
      name: string;
      preview_url?: string;
      artists: Array<{ id: string; name: string }>;
      album: { id: string; name: string; images: Array<{ url: string }> };
    }>;
  };
}
