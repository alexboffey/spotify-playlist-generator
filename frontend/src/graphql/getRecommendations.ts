import gql from "graphql-tag";

export const GET_RECOMMENDATIONS_QUERY = gql`
  query GET_RECOMMENDATIONS_QUERY($seed_artists: String!) {
    getRecommendations(seed_artists: $seed_artists) {
      tracks {
        id
        name
        artists {
          id
          name
        }
        album {
          id
          name
        }
      }
    }
  }
`;

export interface IGetRecommendationsQuery {
  getRecommendations: {
    tracks: Array<{
      id: string;
      name: string;
      artists: { id: string; name: string };
      album: { id: string; name: string };
    }>;
  };
}
