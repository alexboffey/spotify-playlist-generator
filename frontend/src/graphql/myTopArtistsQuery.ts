import gql from "graphql-tag";

export const MY_TOP_ARTISTS_QUERY = gql`
  query MY_TOP_ARTISTS_QUERY(
    $limit: Int = 20
    $time_range: TimeRange = long_term
  ) {
    myTopArtists(limit: $limit, time_range: $time_range) {
      items {
        id
      }
    }
  }
`;

export interface IMyTopArtistsQuery {
  myTopArtists: {
    items: Array<{ id: string }>;
  };
}
