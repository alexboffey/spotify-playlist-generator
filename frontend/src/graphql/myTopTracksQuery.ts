import gql from "graphql-tag";

export const MY_TOP_TRACKS_QUERY = gql`
  query {
    myTopTracks(limit: 5, time_range: short_term) {
      items {
        name
        id
        href
        album {
          name
        }
      }
      next
      previous
      total
      limit
      href
    }
  }
`;

export interface MyTopTracksData {
  myTopTracks: {
    items: Array<{ name: string; id: string; href: string }>;
    next: string;
    previous: string;
    total: number;
    limit: number;
    href: string;
  };
}
