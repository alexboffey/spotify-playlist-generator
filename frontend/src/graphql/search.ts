import gql from "graphql-tag";

export const SEARCH_ARTISTS_QUERY = gql`
  query SEARCH_ARTISTS_QUERY(
    $query: String!
    $type: SearchType = artist
    $limit: Int
  ) {
    search(query: $query, type: $type, limit: $limit) {
      artists {
        items {
          name
        }
      }
    }
  }
`;

export const SEARCH_TRACKS_QUERY = gql`
  query SEARCH_ARTISTS_QUERY(
    $query: String!
    $type: SearchType = track
    $limit: Int
  ) {
    search(query: $query, type: $type, limit: $limit) {
      tracks {
        items {
          name
        }
      }
    }
  }
`;

export interface ISearchQuery {
  search: string;
  type: string;
}
