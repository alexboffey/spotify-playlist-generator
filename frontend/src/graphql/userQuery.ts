import gql from "graphql-tag"

export const USER_QUERY = gql`
  query {
    me {
      id
      name
      email
      spotifyId
    }
  }
`;

export interface IUserQuery {
  me: {
    id: string;
    name: string;
    email: string;
    spotifyId: string;
  };
}
