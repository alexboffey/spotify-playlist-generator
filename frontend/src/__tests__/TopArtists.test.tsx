import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";

import { MY_TOP_ARTISTS_QUERY } from "../graphql/myTopArtistsQuery";
import TopArtists from "../components/TopArtists";

const mocks = [
  {
    request: {
      query: MY_TOP_ARTISTS_QUERY,
      variables: { limit: 20, time_range: "short_term" }
    },
    result: {
      data: {
        myTopArtists: {
          __typename: "test",
          items: [
            {
              id: "someId",
              name: "someName",
              genres: ["genreA", "genreB"],
              images: [{ url: "someImageUrl", __typename: "test" }],
              __typename: "test"
            }
          ]
        }
      }
    }
  }
];

describe("<TopArtists />", () => {
  it("Queries, renders data and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TopArtists />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    // Data assertions
    expect(
      wrapper
        .find(".ant-list-item-meta-title")
        .first()
        .text()
    ).toBe("someName");

    // Snapshot assertion
    expect(
      toJSON(wrapper.find("div.ant-list.ant-list-split"))
    ).toMatchSnapshot();
  });
});
