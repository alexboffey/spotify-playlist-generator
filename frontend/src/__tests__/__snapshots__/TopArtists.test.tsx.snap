// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopArtists /> Queries, renders data and matches snapshot 1`] = `
<div
  className="ant-list ant-list-split"
>
  <Spin
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
        key="container"
      >
        <Item
          key="someId"
        >
          <li
            className="ant-list-item"
          >
            <Meta
              avatar={
                <Avatar
                  shape="square"
                  size="large"
                  src="someImageUrl"
                />
              }
              description="Genre A, Genre B"
              title="someName"
            >
              <div
                className="ant-list-item-meta"
              >
                <div
                  className="ant-list-item-meta-avatar"
                >
                  <Avatar
                    shape="square"
                    size="large"
                    src="someImageUrl"
                  >
                    <span
                      className="ant-avatar ant-avatar-lg ant-avatar-square ant-avatar-image"
                      style={Object {}}
                    >
                      <img
                        onError={[Function]}
                        src="someImageUrl"
                      />
                    </span>
                  </Avatar>
                </div>
                <div
                  className="ant-list-item-meta-content"
                >
                  <h4
                    className="ant-list-item-meta-title"
                  >
                    someName
                  </h4>
                  <div
                    className="ant-list-item-meta-description"
                  >
                    Genre A, Genre B
                  </div>
                </div>
              </div>
            </Meta>
          </li>
        </Item>
      </div>
    </div>
  </Spin>
</div>
`;
